### PR TITLE
created DiverseSelector class and tests

### PR DESCRIPTION
**Pull request by RichRick1**
1. Code a DiverseSelector class
2. Run tests for all possible choices for clustering algorithm

### CODE REVIEW

**Comment by RichRick1** [comment]
@fwmeng88 any suggestions on how to make it better?

**Comment by RichRick1** [comment]
@fwmeng88 seems like all checks are passed
Any suggestions about the structure of the code?

**Comment by fwmeng88** [comment]
I will take a close look and get it back to you. Thanks for the nice work! @RichRick1

File: DiverseSelector/test/test_ClusteringBased.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, see <http://www.gnu.org/licenses/>
+#
+# --
+
+"""Testing for the dissimilarity-based selection algorithms."""
```
Comment (fwmeng88) [inline DiverseSelector/test/test_ClusteringBased.py:24]: Should be **Testing for the clustering-based selection algorithms.**.

File: DiverseSelector/test/test_ClusteringBased.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                                                    num_clusters=3,
+                                                    clustering_method="k-means",
+                                                    metric='euclidian',
+                                                    arr_dist=arr_dist)
+    selector.features = coords
+    selector.cluster()
+    selected_ids = selector.select()
+    labels_seleted = np.array([class_labels[elem] for elem in selected_ids])
+    for i in range(3):
+        assert len(labels_seleted[labels_seleted == i]) == 4
```
Comment (fwmeng88) [inline DiverseSelector/test/test_ClusteringBased.py:50]: I am not quite sure what the coding is checking here. I would assume we should check

- if  the selected label is correct
- if the number of selected data points matches the requirement

File: DiverseSelector/test/test_ClusteringBased.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, see <http://www.gnu.org/licenses/>
+#
+# --
+
+"""Testing for the dissimilarity-based selection algorithms."""
+
```
Comment (fwmeng88) [inline DiverseSelector/test/test_ClusteringBased.py:25]: The file name should be *test_clustering_based.py*. This is a convention not to use capitalized letters in file names.

**Review by fwmeng88 — CHANGES_REQUESTED** [review]
Generally speaking, it's very good. Only a few minor comments.

File: DiverseSelector/test/test_ClusteringBased.py (hunk 2 above)
Comment (RichRick1) [reply DiverseSelector/test/test_ClusteringBased.py:50]: I checked that from every cluster were selected equal amount of points(molecules). So, there are 3 clusters and 12 molecules needs to be selected. In this test, in particular, it is checked if 4 points were selected from every cluster.

File: DiverseSelector/test/test_ClusteringBased.py (hunk 1 above)
Comment (RichRick1) [reply DiverseSelector/test/test_ClusteringBased.py:24]: Fixed

File: DiverseSelector/clustering_based.py [hunk 4]
```diff
@@ -22,3 +22,126 @@
...
+                 metric="Tanimoto",
+                 feature_file=None,
+                 feature_type=None,
+                 mol_file=None,
+                 output=None,
+                 random_seed=None,
+                 arr_dist=None,
+                 **kwargs
+                 ):
+        """Base class for clustering based subset selection."""
```
[the remark this answers is not in the document]
Comment (RichRick1) [reply DiverseSelector/clustering_based.py:54]: Fixed

**Comment by fwmeng88** [comment]
This is good at this moment and I am going to merge it. Thanks for the work. @RichRick1

**Pull request merged**